File: src/RNetPi.Core/RNet/RequestDataPacket.cs
using System;

namespace RNetPi.Core.RNet;

/// <summary>
/// Packet for requesting data from the RNet system
/// </summary>
public class RequestDataPacket : DataPacket
{
    public RequestDataPacket(byte[] targetPath)
    {
        MessageType = 0x00;
        TargetPath = targetPath ?? throw new ArgumentNullException(nameof(targetPath));
        SourcePath = Array.Empty<byte>();
        PacketNumber = 0;
        PacketCount = 1;
        Data = Array.Empty<byte>();
    }

    public override bool RequiresHandshake()
    {
        return true;
    }

    public override bool CausesResponseWithHandshake()
    {
        return true;
    }

    /// <summary>
    /// Creates a request for zone information
    /// </summary>
    public static RequestDataPacket CreateZoneInfoRequest(byte controllerID, byte zoneID)
    {
        var targetPath = new byte[] { 0x02, 0x00, controllerID, 0x07 };
        var packet = new RequestDataPacket(targetPath)
        {
            TargetControllerID = controllerID,
            TargetZoneID = zoneID
        };
        return packet;
    }

    /// <summary>
    /// Creates a request for zone power status
    /// </summary>
    public static RequestDataPacket CreateZonePowerRequest(byte controllerID, byte zoneID)
    {
        var targetPath = new byte[] { 0x02, 0x00, controllerID, 0x06 };
        var packet = new RequestDataPacket(targetPath)
        {
            TargetControllerID = controllerID,
            TargetZoneID = zoneID
        };
        return packet;
    }

    /// <summary>
    /// Creates a request for zone source status
    /// </summary>
    public static RequestDataPacket CreateZoneSourceRequest(byte controllerID, byte zoneID)
    {
        var targetPath = new byte[] { 0x02, 0x00, controllerID, 0x02 };
        var packet = new RequestDataPacket(targetPath)
        {
            TargetControllerID = controllerID,
            TargetZoneID = zoneID
        };
        return packet;
    }

    /// <summary>
    /// Creates a request for zone volume status
    /// </summary>
    public static RequestDataPacket CreateZoneVolumeRequest(byte controllerID, byte zoneID)
    {
        var targetPath = new byte[] { 0x02, 0x00, controllerID, 0x01 };
        var packet = new RequestDataPacket(targetPath)
        {
            TargetControllerID = controllerID,
            TargetZoneID = zoneID
        };
        return packet;
    }
}

File: src_js/rnet/packets/RequestDataPacket.js
const RNetPacket = require("./RNetPacket");
const SmartBuffer = require("smart-buffer").SmartBuffer;

class RequestDataPacket extends RNetPacket {
    constructor(controllerID, zoneID, dataType) {
        super();

        this.targetControllerID = controllerID;
        this.messageType = 0x01;
        this.sourcePath = [];
        this.targetPath = [
            0x02,
            0x00,
            zoneID,
            dataType
        ]
    }

    getMessageBody() {
        const buffer = new SmartBuffer();

        buffer.writeUInt8(this.targetPath.length);
        for (let i = 0; i < this.targetPath.length; i++) {
            buffer.writeUInt8(this.targetPath[i])
        }

        buffer.writeUInt8(this.sourcePath.length);
        for (var i = 0; i < this.sourcePath.length; i++) {
            buffer.writeUInt8(this.sourcePath[i])
        }

        // Request Data Type
        // NOTE Assuming this is always 0;
        buffer.writeUInt8(0x00);

        return buffer.toBuffer();
    }

    causesResponseWithHandshake() {
        return true;
    }
}

RequestDataPacket.DATA_TYPE = {
    ZONE_INFO: 0x07,
    ZONE_POWER: 0x06,
    ZONE_SOURCE: 0x02,
    ZONE_VOLUME: 0x01
}

module.exports = RequestDataPacket;
